docs/GroupDefs.hpp

```diff
@@ -159,155 +164,485 @@
+
+Again a reference to the DataBox is returned, while the termination `bool` and
+next action `size_t` are returned by value. The metafunction call
+`tmpl::index_of<ActionList, iterate_increment_int0>::%value` returns a `size_t`
+whose value is that of the action `iterate_increment_int0` in the `action_list`.
+The indexing of actions in the `action_list` starts at `0`.
+
+Iterable actions are invoked as part of the algorithm and so the only way
+to request they be invoked is by having the algorithm run on the parallel
+component. The algorithm can be explicitly triggered by call the
+`perform_algorithm` method:
+
+\snippet Test_AlgorithmCore.cpp perform_algorithm
+
+The algorithm is also evaluated by calling the `receive_data` function, either
+on an entire array or singleton (this does a broadcast), or an on individual
+element of the array. Here is an example of a broadcast call:
+
+\snippet Test_AlgorithmParallel.cpp broadcast_to_group
+
+and of calling individual elements:
+
+\snippet Test_AlgorithmParallel.cpp call_on_indexed_array
+
+The `receive_data` function always takes a `ReceiveTag`, which is set in the
+actions `inbox_tags` type alias as described above.  The first argument is the
+temporal identifier, and the second is the data to be sent.
+
+#### 3. Reduction %Actions
+
+Finally, there are reduction actions which are used when reducing data over an
+array. For example, you may want to know the sum of a `int` from every
+element in the array. You can do this by using Charm++'s built-in reduction
+function for summing `int`s as follows:
+
+\snippet Test_AlgorithmReduction.cpp contribute_to_reduction_example
+
+in the array. This reduces over the parallel component
+`ArrayParallelComponent<Metavariables>`, reduces to the parallel component
+`SingletonParallelComponent<Metavariables>`, and calls the action
+`singleton_reduce_sum_int`. The `CkReduction::sum_int` argument tells Charm++
+that you want to sum integers instead of some other type or operation. The
+reduction action is:
+
+\snippet Test_AlgorithmReduction.cpp singleton_reduce_sum_int
+
+As you can see, the last argument to the `apply` function is of type `int`, and
+is the reduced value.
+
+You can also broadcast the result back to an array, even yourself. For example,
+
+\snippet Test_AlgorithmReduction.cpp contribute_to_broadcast_reduction
+
+It is often necessary to reduce custom data types, such as `std::vector` or
+`std::unordered_map`. Charm++ support such custom reductions, and so does our
+layer on top of Charm++.
+Custom reductions require one additional step to calling
+`contribute_to_reduction`, which is writing a reduction function to reduce the
+custom data. We provide a generic type that can be used in custom reductions,
+`Parallel::ReductionData`, which takes as template parameters the types it
+holds, similar to `std::tuple`. An example of a custom reduction function is:
+
+\snippet Test_AlgorithmReduction.cpp custom_reduce_function
```

Comment:
Perhaps `reduce_custom_data` would be a better name.

Reply:
This comment has nothing to do with the documentation. The snippet name is largely irrelevant, and the current one I think is actually clearer than what you suggested. The rendered function has the name `reduce_reduction_data`.

Reply:
_Nothing to do with documentation_ is a bit of an overstatement, it’s rendered in the documentation after all, and tests serve also an illustrative purpose. Then take this comment, and the previous, as a suggestion to adjust the Test code.

Reply:
Not really, the function name you are talking about already has the name `reduce_reduction_data`, are you asking that be changed? I'm not changing the snippet name.

Reply:
I'm confused by this discussion.  Are you two sure you're talking about the same thing?

Reply:
Yeah the snippet name is irrelevant :D I simply suggested to make this snippet‘s content, which is pulled from a test, more expressive. In fact, if you prefer the naming of this particular function to remain as it is I would not object, but I think the snippet `singleton_reduce_sum_int` could be improved, as I commented above.

Reply:
Okay, so at least we agree on that :) I've changed the `single_reduce_sum_int` (not pushed yet). I'm fine changing the name of the function as well, was just confused.

Reply:
After thinking about it more, I like `reduce_reduction_data` since the type it's reducing over is a `ReductionData`